## Phase1: Exploring the dataset
Prior to analyzing, we'll understand the structure of data , its shape , types of columns as well as if any values are missing 

In [11]:
import pandas as pd
schools=pd.read_csv("data/schools.csv")
print(schools.shape)
schools.head(20)

(375, 7)


,school_name,borough,building_code,average_math,average_reading,average_writing,percent_tested
0,"New Explorations into Science, Technology and ...",Manhattan,M022,657,601,601,NaN
1,Essex Street Academy,Manhattan,M445,395,411,387,78.9
2,Lower Manhattan Arts Academy,Manhattan,M445,418,428,415,65.1
3,High School for Dual Language and Asian Studies,Manhattan,M445,613,453,463,95.9
4,Henry Street School for International Studies,Manhattan,M056,410,406,381,59.7
5,Bard High School Early College,Manhattan,M097,634,641,639,70.8
6,Urban Assembly Academy of Government and Law,Manhattan,M445,389,395,381,80.8
7,Marta Valle High School,Manhattan,M025,438,413,394,35.6
8,University Neighborhood High School,Manhattan,M446,437,355,352,69.9
9,New Design High School,Manhattan,M445,381,396,372,73.7


In [12]:
schools.info()
schools.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 375 entries, 0 to 374
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   school_name      375 non-null    object 
 1   borough          375 non-null    object 
 2   building_code    375 non-null    object 
 3   average_math     375 non-null    int64  
 4   average_reading  375 non-null    int64  
 5   average_writing  375 non-null    int64  
 6   percent_tested   355 non-null    float64
dtypes: float64(1), int64(3), object(3)
memory usage: 20.6+ KB


,average_math,average_reading,average_writing,percent_tested
count,375.000000,375.000000,375.000000,355.000000
mean,432.944000,424.504000,418.458667,64.976338
std,71.952373,61.881069,64.548599,18.747634
min,317.000000,302.000000,284.000000,18.500000
25%,386.000000,386.000000,382.000000,50.950000
50%,415.000000,413.000000,403.000000,64.800000
75%,458.500000,445.000000,437.500000,79.600000
max,754.000000,697.000000,693.000000,100.000000


In [13]:
schools.isnull().sum()

school_name         0
borough             0
building_code       0
average_math        0
average_reading     0
average_writing     0
percent_tested     20
dtype: int64

**Findings**
the dataset has 375 rows and 7 columns, all columns are not null except `percent_tested` which has 20 missing values found via `df.isnull().sum()` 


## Phase 2: Handling missing values
The `percent_tested` column has 20 missing values out of 375 schools (~5.3%). Not huge yet not negligible.
**Decision:** we'll keep all 375 rows in thedataset, but filter the missing `percent_tested` rows once needed for analysis

### Total SAT score per school

In [ ]:
schools["total_SAT"]=schools["average_math"]+schools["average_reading"]+schools["average_writing"]
schools.head(3)

,school_name,borough,building_code,average_math,average_reading,average_writing,percent_tested,total_SAT
0,"New Explorations into Science, Technology and ...",Manhattan,M022,657,601,601,NaN,1859
1,Essex Street Academy,Manhattan,M445,395,411,387,78.9,1193
2,Lower Manhattan Arts Academy,Manhattan,M445,418,428,415,65.1,1261
3,High School for Dual Language and Asian Studies,Manhattan,M445,613,453,463,95.9,1529
4,Henry Street School for International Studies,Manhattan,M056,410,406,381,59.7,1197
5,Bard High School Early College,Manhattan,M097,634,641,639,70.8,1914
6,Urban Assembly Academy of Government and Law,Manhattan,M445,389,395,381,80.8,1165
7,Marta Valle High School,Manhattan,M025,438,413,394,35.6,1245
8,University Neighborhood High School,Manhattan,M446,437,355,352,69.9,1144
9,New Design High School,Manhattan,M445,381,396,372,73.7,1149


### Best math score

In [24]:
math_max=schools["average_math"].max()
schools.loc[schools["average_math"] == math_max, ["school_name","average_math"]
            ]

,school_name,average_math
88,Stuyvesant High School,754


### Top 10 schools by total SAT

In [29]:
schools.sort_values( by="total_SAT", ascending=False).head(10)[["school_name","total_SAT"]]

,school_name,total_SAT
88,Stuyvesant High School,2144
170,Bronx High School of Science,2041
93,Staten Island Technical High School,2041
174,High School of American Studies at Lehman College,2013
333,Townsend Harris High School,1981
365,Queens High School for the Sciences at York Co...,1947
5,Bard High School Early College,1914
280,Brooklyn Technical High School,1896
45,Eleanor Roosevelt High School,1889
68,"High School for Mathematics, Science, and Engi...",1889


### Borough with the most schools in the top 10

In [32]:
top_10=schools.sort_values(by="total_SAT" , ascending=False).head(10)
top_10["borough"].value_counts()

Manhattan        4
Bronx            2
Queens           2
Staten Island    1
Brooklyn         1
Name: borough, dtype: int64

The borough with the most schools in the top 10 by total SAT score is **Manhattan**, with **4** schools.

### Borough with the highest and lowest average total SAT

In [42]:
schools.groupby("borough")["total_SAT"].mean().sort_values(ascending=False)

borough
Staten Island    1439.000000
Queens           1345.478261
Manhattan        1340.134831
Brooklyn         1230.256881
Bronx            1202.724490
Name: total_SAT, dtype: float64

### Borough summary: school count, mean, and std dev of total SAT, sorted by std dev

In [44]:
schools.groupby("borough")["total_SAT"].agg(["count", "mean", "std"]).sort_values(by="std", ascending=False)

,count,mean,std
borough,,,
Manhattan,89,1340.134831,230.294140
Staten Island,10,1439.000000,222.303596
Queens,69,1345.478261,195.252899
Brooklyn,109,1230.256881,154.868427
Bronx,98,1202.724490,150.393901


### Understanding reset_index() before and after a groupby

In [45]:
schools.groupby("borough")["total_SAT"].agg(["count", "mean", "std"]).sort_values(by="std", ascending=False).reset_index()

,borough,count,mean,std
0,Manhattan,89,1340.134831,230.294140
1,Staten Island,10,1439.000000,222.303596
2,Queens,69,1345.478261,195.252899
3,Brooklyn,109,1230.256881,154.868427
4,Bronx,98,1202.724490,150.393901
